_source/_posts/2020-06-29-aspnet-core-mongodb.md

```diff
@@ -0,0 +1,513 @@
+  padding: .5rem;
+}
+
+.idea > div {
+  padding: .5rem;
+}
+
+a.create {
+  display: block;
+  width: 100%;
+  text-align: right;
+}
+
+.idea-form > * {
+  display: block;
+}
+
+.idea-form > textarea {
+  height: 5rem;
+  width: 100%;
+}
+```
+
+## Add Authentication to the .NET Core Application
+
+Lastly, you'll need to add authentication using Okta so that the application can enforce the rule that only users who created a submission can edit or delete those submissions. Start by creating an application in Okta—go to your Okta dashboard, click on **Applications** in the top menu and then click the green **Add Application** button. Choose **Web** and click **Next**.
+
+In the **Application Settings**, name the application "SuggestMe" then change the **Base URIs** to `http://localhost:5000/`, the **Login redirect URIs** to `http://localhost:5000/signin-oidc` and the **Logout redirect URIs** to `http://localhost:5000/signout-callback-oidc`. All the other default values should be fine.
+
+{% img blog/aspnet-core-mongodb/okta-app-settings.png alt:"Okta App Settings" width:"800" %}{: .center-image }
+
+When you click **Done**, you should be directed to the application settings page. On the **General** tab at the bottom, you'll see a box with a **ClientId** and **ClientSecret**. Copy those to a new section in `appsettings.json` below the database settings you added earlier:
+
+```json
+"Okta": {
+    "ClientId": "{yourClientId}",
+    "ClientSecret": "{yourClientSecret}",
+    "Domain": "{yourOktaDomain}",
+    "PostLogoutRedirectUri": "http://localhost:5000"
+},
+```
+
+You can find your Okta domain URL on the top right of your Okta dashboard.
+
+{% img blog/aspnet-core-mongodb/okta-org-url.png alt:"Okta Org URL" width:"800" %}{: .center-image }
+
+Next, add a file called `AccountController.cs` to the `Controllers` folder like so:
+
+```cs
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.AspNetCore.Mvc;
+
+namespace DockerPipelineExample.Controllers
+{
+  public class AccountController : Controller
+  {
+    public IActionResult Login()
+    {
+      if(!HttpContext.User.Identity.IsAuthenticated)
+      {
+        return Challenge(OpenIdConnectDefaults.AuthenticationScheme);
+      }
+      return RedirectToAction("Index", "Home");
+    }
+
+    public IActionResult Logout()
+    {
+      return new SignOutResult(new[]
+      {
+        OpenIdConnectDefaults.AuthenticationScheme,
+        CookieAuthenticationDefaults.AuthenticationScheme
+      });
+    }
+  }
+}
+```
+
+This adds actions for logging in and logging out that can then be used by the application to facilitate authentication with Okta. 
+
+Finally, you'll add navigation so that users can get to all this cool stuff! In the `Views/Shared` folder add a file called `_LoginPartial.cshtml`. This will be the partial view that contains navigation and display elements for authentication routes:
+
+```html
```

Comment:
This code block has 2 spaces instead of 4. Make it 4 to be like the rest of the code blocks.